### PR TITLE
Ensure IAccessible2 remains available in Firefox after restarting NVDA on Windows 10 Creaters update, register ISimpleDOM proxy to allow math in Google Chrome without Firefox installed, and Allow the user to get back a virtualBuffer when restarting NVDA with a Firefox document in focus

### LINKED ISSUE #7269: Browse mode in Mozilla Gecko products no longer can be used after restarting NVDA due to IAccessible2 interface registration issues
### Steps to reproduce:
This can only be reproduced on a small amount of machines running Windows 10 Creaters Update.
While NVDA is running:
* Start Firefox and navigate a page with NVDA browse mode
* Restart NVDA
* Move back to Firefox and try and navigate the page.

Browse mode  will not be available due to NVDA not being able to queryService to IAccessibleApplication to fetch the Gecko version.

### LINKED ISSUE #7308: Build and register ISimpleDOM proxy ourselves (AKA fix broken math in Firefox/Chrome on some systems)
### Steps to reproduce:
1. Open this URL in Firefox or Chrome: `data:text/html,<math><mi>x</mi></math>`
2. Use the cursor keys to read the document.

### Expected behavior:
"x" should be reported.

### Actual behavior:
On some systems/builds of software, nothing is reported and the following is logged:

> ```
> DEBUGWARNING - NVDAObjects.IAccessible.ia2Web.Math._get_mathMl (11:45:38.736):
> Error retrieving math. Not supported in this browser or ISimpleDOM COM proxy not registered.
> Traceback (most recent call last):
>   File "NVDAObjects\IAccessible\ia2Web.pyc", line 63, in _get_mathMl
>   File "comtypes\__init__.pyc", line 1078, in QueryInterface
> COMError: (-2147467262, 'No such interface supported', (None, None, None, 0, None))
> ```

### Additional info:
This occurs when the 32 bit and/or 64 bit ISimpleDOM COM proxy is not registered. (For 64 bit browsers, we need both.) This can happen for several reasons:

1. 64 bit builds of Firefox don't install the 32 bit proxy. This will only become more prevalent as more users migrate to 64 bit builds.
2. I suspect (but am not certain) that Chrome doesn't ever install the proxy, regardless of bitness.
3. If you uninstall a Mozilla app while other Mozilla apps are still installed, that might unregister the proxy.
4. Possibly other obscure edge cases I haven't thought of.

The key point is that we simply cannot rely on this being installed correctly. Instead, we should build it and register it ourselves using CoRegisterClassObject/CoRegisterPSClsid as we do for IAccessible2 (see installIA2Support in nvdaHelper/remote/IA2Support.cpp).

P2 because this breaks math support for Firefox and Chrome for an increasing number of users.

### LINKED ISSUE #7563: other apps may crash or cause errors when an NVDA installation is updated
### Steps to reproduce:
1. Install [NVDA 2017.3](https://www.nvaccess.org/files/nvda/releases/2017.3/nvda_2017.3) and reboot.
 2. With NVDA running, open apps such as Firefox or Mumble.
 3. Install [this NVDA Next snapshot](https://ci.appveyor.com/api/buildjobs/nl9hyd708emuia09/artifacts/output/nvda_snapshot_next-14383,d06cd200.exe)
4. Once NVDA has installed and restarted itself, Start alt+tabbing around the system.
 
### Expected behavior:
The apps should not crash.

### Actual behavior:
One or more of the apps crash.

### System configuration:
Windows version: 10.0.16278 64-bit

#### We forget to unhookWinEvent in NVDAHelperRemote
NVDAHelperRemote injects into applications with foreground and focus winEvents. In inprocMgrThreadFunc, it then registers another winEvent limited to that process (inproc_winEventCallback) that listens for all possible winEvents. On termination, the injection winEvent is unregistered, but the process-specific winEvent is not. This leaves NVDAHelperRemote.dll in the process for ever. It will also respond to winEvents even though nvdaHelperRemote is terminated which may cause instability or a crash. It is necessary that inproc_winEventCallback be unregistered before inprocMgrThreadFunc exits.

#### Accidentally execute invalid memory
It is possible to accidentally cause Windows to execute incorrect memory in an in-context winEvent hook and thereby crash the injected process, when using a valid dll and a valid executable function pointer from that dll.
Specifically:
Create 2 dlls. Each with a winEventCallback function. However, add extra code to dll1's dllMain so that it does an extra addref (GetModuleHandleEx) on process attach so that it never unloads.
Then:
A. Rename dll1 to tempPath
A. Load dll 1 via tempPath into process1
B. Call SetWinEventHook giving it the handle of dll1, and the function pointer of a WINEVENTCALLBACK within dll1, make it listen for events in process2, in-context.
C. Fire a winEvent in process2, so that Windows injects dll1 and executes the winEventCallback.
D. Unhook dll1's winEventCallback in process1
E. Rename dll2 to tempPath (replacing dll1)
F. Load dll2 via tempPath into process1
G. Call SetWinEventHook giving it the handle of dll2, and the function pointer of a WINEVENTCALLBACK within dll2, make it listen for events in process2, in-context.
H. Fire a winEvent in process2, so that Windows *should* inject dll2 and execute the winEventCallback.
However:
As dll1 and dll2 both have exactly the same path, and dll1 is still loaded, Windows does not bother loading dll2. Rather it uses dll1's memory address to calculate the physical location of the winEventCalback. But, as this was originally from dll2, it hits random memory!
In other words, It seems that the only information sent from process1 to process2 is the path of the dll, and the offset of the winEventCallback (I.e. winEventCallback minus dll base address).  Then in process2, it does a simple LoadLibrary(path) and calculates dllHandle plus winEventCallback offset.

#### Practical outcome
Normally when updating an NVDA installation, as nvdaHelperRemote.dll did not change its layout, things go okay. However, if nvdaHelperRemote is changed in any dramatic way (such as a segment being removed or code moved around such as with the mentioned Next snapshot) then if the old nvdaHelperRemote takes a long time to unload from an app, and in deed it is still there when the new NVDA tries to inject, Windows will refuse to load in the new copy as it thinks it already has it, and then executes from the wrong memory location.
We specifically hit this problem when implementing pr #7535 

#### Suggested fix
We should place all NVDA dlls in a version-specific directory, so that the path to any NVDA dll changes for each version.

### PR DESCRIPTION
**Pull request by michaelDCurran**
### Link to issue number:
Fixes #7269 
Fixes #7308
Partial fix for #5758
Fixes #7563 

### Summary of the issue:

Originally some users were complaining that when they restarted NVDA while Firefox was in focus, browse mode was no longer available. Alt+tabbing away and back again would also not correct the situation.
Specifically issue #5758 covers this original case. Technically, when NVDA restarts and Firefox is already in focus, There is no focus event for Firefox and therefore NVDA's in-proces injection code is not triggered for the Firefox process. Therefore we have no binding handle  for creating a virtualBuffer. However, as VirtualBuffer.prepare has already set shouldPrepare to false by the time it reaches this error, it is impossible for the virtualBuffer to ever work again, even if a binding handle does become available by shifting focus away and back again.

There was a second issue (#7269) that started appearing in Windows insider builds  approaching the Windows 10 Creaters Update.  In this situation, restarting NVDA and then trying to use Firefox would result in NVDA not being able to ever QueryService to IAccessibleApplication to check the Gecko version number, and therefore NVDA would never provide browse mode. Restarting NVDA would not help. All that could be done was to close and re-open Firefox. There seemed to be some kind of COM proxy stub issue where if a request was made to marshal an interface that had just been deregistered with CoRevokeClassObject, then once that interface was again successfully registered, marshalling would be for ever still broken in the server process.

Finally, issue #7308 talks about not being able to read / navigate math content in Google chrome. The main reason for this is that Google Chrome does not register ISimpleDOM interfaces which are used between chrome and NVDA to fetch the math content. What confuses the situation is that if the user did have Firefox installed, math content would work in Google chrome as the interfaces were available via Firefox's registration.

 
### Description of how this pull request fixes the issue:
#5758 is partially fixed by simply returning early in VirtaulBuffer.prepare if there is no binding handle yet. This means that on future focus changes, the VirtaulBuffer will again try to prepare. Once there is a binding handle, this will succeeed. In short, restarting NVDA while Firefox is in focus will still result in no browse mode at first, but the user can easily shift focus away and back again to get Browse mode.

#7269 and #7308 together are fixed by a substantial re-write to NVDAHelperRemote's COM proxy registration code.
Firstly, the code has been abstracted so that it is easy to register other COM proxies other than just IAccessible2. Now both IAccessible2 and ISimpleDOM are both built, and registered in all processes we inject into if COM is initialized. The abstraction also allows us to fetch all the interfaces from the dll directly, rather than having to hardcode them.
The COM proxy dlls now contain assembly manifests which aide in us getting COM to load the class objects itself, while manging the life time of the dll, saving us from having to free it on NVDA shutdown, which in 99% of cases never worked.
And, most importantly the specific problem introduced by Windows 10 creaters update where CoRevokeClassObject would break future registration was fixed by pointing the interfaces to the standard marshaller with CoRegisterPSClsid before their class object was revoked. The standard marshaller does not support the interfaces of course, but at least they still point to a valid class object. Microsoft really needs a CoUnregisterPSClsid.
 
Early iterations of this PR also showed up issue #7563, which this PR now also addresses by moving all NVDA dlls into a version-specific directory to stop accidental loading or reuse of a dll from a previous version which could cause a crash.

### Testing performed:
Multiple restarts of NVDA both using control+alt+n, and the NVDA+q quit dialog (choosing restart), on both Windows 10 Creaters update and Windows 7 64 bit, while Mozilla Firefox is in focus.

### Known issues with pull request:
Issue #5758 is not fully fixed as the user must still alt+tab away and back again to get Browse mode in Firefox when restarting NVDA with Firefox in focus.

### Change log entry:
Bug fixes:
* On windows 10 Creaters Update, NVDA can again access Firefox after a restart of NVDA. (#7269)
* When restarting NVDA with Mozilla Firefox in focus, browse mode  will again be available, though you may need to alt+tab away and back again. (#5758)
* It is now possible to access math content in Google Chrome on a system with out Mozilla Firefox installed. (#7308)
* The Operating System and other applications should be more stable directly after installing NVDA before rebooting, as compaired with installs of previous NVDA versions. (#7563)

### CODE REVIEW

**Comment by leonardder** [comment]
I assume this doesn't contribute anything to a fix of #7311?

**Comment by michaelDCurran** [comment]
This won't directly affect the problems in that issue. But all of these 
issues are similar in that COM marshalling is affected. We are planning 
to include some kind of utility in NVDA to fix issues such as #7311 and 
perhaps having it run in the NVDA installer also.

File: nvdaHelper/remote/COMProxyRegistration.cpp [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+				LOG_DEBUG(L"Backed up existing clsid for interface "<<name);
+			}
+			if((res=CoRegisterPSClsid(iid,regClsid))!=S_OK) {
+				LOG_ERROR(L"Unable to register interface "<<name<<L" with proxy stub "<<dllPath<<L", code "<<res);
+				continue;
+			}
+		reg->psClsidBackups.push_back({name,iid,clsidBackup});
+		LOG_DEBUG(L"Registered interface "<<name);
+		}
+		++tempInfoPtr;
```
Comment (feerrenrut) [inline nvdaHelper/remote/COMProxyRegistration.cpp]: Is there some docs you can point to on ProxyFileInfo? I'm a little nervous about this code. So `pProxyInfo` points to the start of a memory block containing `ProxyFileInfo*`. This memory block is null terminated. I cant find any documentation on `Dll_GetProxyDllInfo` How do you know the above is the case?

File: nvdaHelper/remote/COMProxyRegistration.cpp [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+	HRESULT res;
+	LOG_DEBUG(L"Unregistering proxy "<<(reg->dllPath));
+	for(auto& backup: reg->psClsidBackups) {
+		if((res=CoRegisterPSClsid(backup.iid,backup.clsid))!=S_OK) {
+			LOG_ERROR(L"Error registering backup PSClsid for interface "<<(backup.name)<<L" from "<<(reg->dllPath)<<L", code "<<res);
+		}
+		LOG_DEBUG(L"Unregistered interface "<<(backup.name));
+	}
+	if((res=CoRevokeClassObject((DWORD)(reg->classObjectRegistrationCookie)))!=S_OK) {
+		LOG_ERROR(L"Error unregistering class object from "<<(reg->dllPath)<<L", code "<<res);
```
Comment (feerrenrut) [inline nvdaHelper/remote/COMProxyRegistration.cpp]: For this (and the other error case on line 156) should the function exit early, why is it ok to continue to `delete reg` and `return true`

File: nvdaHelper/remote/COMProxyRegistration.h [hunk 3]
```diff
@@ -0,0 +1,61 @@
+/*
+This file is a part of the NVDA project.
+URL: http://www.nvda-project.org/
+Copyright 2006-2010 NVDA contributers.
```
Comment (feerrenrut) [inline nvdaHelper/remote/COMProxyRegistration.h]: See copyright headers https://github.com/nvaccess/nvda/wiki/Copyright-headers

**Review by feerrenrut — CHANGES_REQUESTED** [review]

**Review by feerrenrut — APPROVED** [review]

**Comment by leonardder** [comment]
This seems to cause incidental errors, which I've seen in at least Thunderbird and Spotify when focussing the application

```
ERROR - RPC process 12692 (thunderbird.exe) (08:50:15.348):
Thread 12696, build\x86\remote\COMProxyRegistration.cpp, registerCOMProxy, 69:
GetProxyDllInfo function not found in IAccessible2Proxy.dll

ERROR - RPC process 12692 (thunderbird.exe) (08:50:15.361):
Thread 12696, build\x86\remote\ia2Support.cpp, installIA2Support, 59:
Error registering IAccessible2 proxy
```

**Comment by michaelDCurran** [comment]
Was this a source or binary copy? If binary, where exactly did you get it?

That error is pretty much suggesting that a different copy of 
IAccessible2Proxy.dll was loaded instead of the one that actually came 
with that version of NVDA.

Very very strange.

At least, does the issue go away if you reboot?

**Comment by leonardder** [comment]
This was a binary copy, Version next-14364,d19efb8d. Note that I hadn't 
yet rebooted, and I assume a reboot is likely to fix this. I'll report 
back if the problem still persists.

**Comment by michaelDCurran** [comment]
This PR also fixes #7563,  which was uncovered by other parts of this PR.

**Review by feerrenrut — APPROVED** [review]

**Pull request merged**